File: java/low-level-design/src/com/educosys/parkinglot/ParkingSpot.java
package com.educosys.parkinglot;

public class ParkingSpot {
    private int spotNo;
    private Vehicle vehicle;
    private VehicleType type;

    public ParkingSpot() {}

    public ParkingSpot(int spotNo, VehicleType type ) {
        this.spotNo = spotNo;
        this.type = type;
    }

    public boolean isAvailable(){
        return vehicle == null;
    }

    public void parkVehicle(Vehicle vehicle)
    {
        if(isAvailable() && type == vehicle.getType())
            this.vehicle = vehicle;

        else
            System.out.println("### Parking spot is not empty");
    }

    public void unparkVehicle()
    {
        this.vehicle = null;
    }

    public int getSpotNo() {
        return spotNo;
    }

    public void setSpotNo(int spotNo) {
        this.spotNo = spotNo;
    }

    public Vehicle getVehicle() {
        return vehicle;
    }

    public void setVehicle(Vehicle vehicle) {
        this.vehicle = vehicle;
    }

    public VehicleType getType() {
        return type;
    }

    public void setType(VehicleType type) {
        this.type = type;
    }
}

File: c++/parking-lot-lld/ParkingSpot.cpp
#include "ParkingSpot.h"
using namespace std;

ParkingSpot::ParkingSpot(int spotNo, VehicleType type)
{
    this->parkedVehicle = NULL;
    this->spotNo = spotNo;
    this->vehicleType = type;
}

bool ParkingSpot::isParkingAvailable(){
    return this->parkedVehicle == NULL;
}

void ParkingSpot::parkVehicle(Vehicle *vehicle)
{
    if(isParkingAvailable() && vehicle->getType() == this->vehicleType)
    {
        this->parkedVehicle = vehicle;
    }
    else{
        cout<<endl<<"Parking Unsuccessful :( by ParkingSpot";
    }
}

void ParkingSpot::unparkVehicle(){
    this->parkedVehicle=NULL;
}

void ParkingSpot::displayParking()
{
    cout<<"Spot No:"<<this->spotNo<<", Availability:"<<(this->isParkingAvailable() ? "Available" : "Unavailable")<<endl;
}

Vehicle* ParkingSpot::getParkedVehicle(){
    return parkedVehicle;
}

VehicleType ParkingSpot::getVehicleType(){
    return vehicleType;
}